# Setup

In [1]:
# import libraries 
import pandas as pd
import numpy as np
import os
# create shortcut path
path = r"C:\Users\HP\Documents\06-2024 Instacart Basket Analysis"
# import data sets
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Data consistency checks on df_ords

In [3]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

In [4]:
# converting order_id, user_id, and order_number to string
df_ords['order_id'] = df_ords['order_id'].astype('str')
df_ords['user_id'] = df_ords['user_id'].astype('str')
df_ords['order_number'] = df_ords['order_number'].astype('str')

In [5]:
# verifying data type change
df_ords.describe()

Unnamed: 0  order_day_of_week  order_hour_of_day  \
count  3.421083e+06       3.421083e+06       3.421083e+06   
mean   1.710541e+06       2.776219e+00       1.345202e+01   
std    9.875817e+05       2.046829e+00       4.226088e+00   
min    0.000000e+00       0.000000e+00       0.000000e+00   
25%    8.552705e+05       1.000000e+00       1.000000e+01   
50%    1.710541e+06       3.000000e+00       1.300000e+01   
75%    2.565812e+06       5.000000e+00       1.600000e+01   
max    3.421082e+06       6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [6]:
df_ords['order_id'].dtype
df_ords['user_id'].dtype
df_ords['order_number'].dtype

dtype('O')

In [7]:
# frequency check - 'order_day_of_week
df_ords['order_day_of_week'].value_counts(dropna = False)

order_day_of_week
0    600905
1    587478
2    467260
5    453368
6    448761
3    436972
4    426339
Name: count, dtype: int64

In [8]:
# frequency check - 'order_hour_of_day'
df_ords['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

In [9]:
# frequency check - 'days_since_prior_order'
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

# Mixed-Type data practice

In [11]:
# create a practice dataframe 
df_test = pd.DataFrame()

In [12]:
# create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [13]:
# check for mixed types 
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


# Missing Values

In [15]:
# finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [16]:
# creating subset of missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [17]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [18]:
# reviewing dimensions of df_prods
df_prods.shape

(49693, 5)

In [19]:
# creating subset of non-missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [20]:
# verifying missing columns have been removed
df_prods_clean.shape

(49677, 5)

# Duplicates

In [22]:
# finding duplicates in df_prods_clean
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
# verifying dimensions
df_prods_clean.shape

(49677, 5)

In [25]:
# removing duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5

### 01. Consistency checks on df_prods_clean_no_dups

In [29]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [30]:
# converting product_id, aisle_id, and department_id to string
df_prods_clean_no_dups.loc[:,'product_id'] = df_prods_clean_no_dups['product_id'].astype('str')
df_prods_clean_no_dups.loc[:,'aisle_id'] = df_prods_clean_no_dups['aisle_id'].astype('str')
df_prods_clean_no_dups.loc[:,'department_id'] = df_prods_clean_no_dups['department_id'].astype('str')

C:\Users\HP\AppData\Local\Temp\ipykernel_24572\1215025390.py:2: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '['1' '2' '3' ... '49686' '49687' '49688']' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df_prods_clean_no_dups.loc[:,'product_id'] = df_prods_clean_no_dups['product_id'].astype('str')
C:\Users\HP\AppData\Local\Temp\ipykernel_24572\1215025390.py:3: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '['61' '104' '94' ... '112' '41' '73']' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df_prods_clean_no_dups.loc[:,'aisle_id'] = df_prods_clean_no_dups['aisle_id'].astype('str')
C:\Users\HP\AppData\Local\Temp\ipykernel_24572\1215025390.py:4: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '['19' 

In [31]:
df_prods_clean_no_dups.describe()

prices
count  49672.000000
mean       9.993282
std      453.615536
min        1.000000
25%        4.100000
50%        7.100000
75%       11.100000
max    99999.000000

In [32]:
df_prods_clean_no_dups.head()

product_id                                       product_name aisle_id  \
0          1                         Chocolate Sandwich Cookies       61   
1          2                                   All-Seasons Salt      104   
2          3               Robust Golden Unsweetened Oolong Tea       94   
3          4  Smart Ones Classic Favorites Mini Rigatoni Wit...       38   
4          5                          Green Chile Anytime Sauce        5   

  department_id  prices  
0            19     5.8  
1            13     9.3  
2             7     4.5  
3             1    10.5  
4            13     4.3

Observations:

If there are missing values in the price column, I will most likely use the median and not the mean as the max is an outlier. Alternatively, I will look up the price on the web and use that.

### 02. Reviewing df_ords

Done previously at beginning of exercise. I do not see any abnormal values for day of week, hour of day and days since prior order. There are no negative values. There are some missing values in the days since prior order, but I would assume that is the user's first order placed so it would make sense those values are missing.

### 03. Mixed-type data in df_ords

In [38]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
else : print ("no mixed-type data")

no mixed-type data


Observation:

There appears to be no mixed-type data in the 'df_ords' dataframe. I updated the code to have an 'if, else' statement.

### 05. Check for missing values in df_ords

In [42]:
# finding missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Observations:

There are 206,209 missing values in the 'days_since_prior_order' column. This is most likely due to user's placing their first order, so it makes sense for the the field to be blank. I do not want to impute or delete these fields as they are relevant to the data, and therefore will keep the data as is.

### 06. Addressing the missing values

To address the missing values, I will create a subset of the dataframe to view the data of the missing values. It would make sense for the 'order_number' to be '1'. I can then check the frequency of the 'order_number' column using the '.value_counts' function.

In [47]:
# create a subset of the missing values
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [48]:
df_nan_ords

Unnamed: 0 order_id user_id order_number  order_day_of_week  \
0                 0  2539329       1            1                  2   
11               11  2168274       2            1                  2   
26               26  1374495       3            1                  1   
39               39  3343014       4            1                  6   
45               45  2717275       5            1                  3   
...             ...      ...     ...          ...                ...   
3420930     3420930   969311  206205            1                  4   
3420934     3420934  3189322  206206            1                  3   
3421002     3421002  2166133  206207            1                  6   
3421019     3421019  2227043  206208            1                  1   
3421069     3421069  3154581  206209            1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [49]:
# frequency check - 'order_number' in df_nan_ords
df_nan_ords['order_number'].value_counts(dropna = False)

order_number
1    206209
Name: count, dtype: int64

In [50]:
# frequency check - 'order_number' in df_ords
df_ords['order_number'].value_counts(dropna = False)

order_number
1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: count, Length: 100, dtype: int64

Observations:

The number of missing values in 'df_nan_ords' is equivalent to 'order_number' = 1 in 'df_ords'. This confirms the missing 'days_since_prior_order' values indicate the user's first order being placed.

### 07. Checking for duplicate values in df_ords

In [54]:
# finding duplicate values
df_dups_ords = df_ords[df_ords.duplicated()]

In [55]:
df_dups_ords.shape

(0, 7)

Observations:

There appears to be no duplicate values in the 'df_ords' dataframe.

In [92]:
# dimensions of df_ords
df_ords.shape

(3421083, 7)

### 09. Exporting updated dataframes

In [60]:
# exporting df_prods_clean_no_dups dataframe as 'products_checked.csv'
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [61]:
# exporting df_ords dataframe as 'orders_checked.csv'
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))